### PR TITLE
Refactor barstools for new versions of things.

### PR DESCRIPTION
**Pull request by grebe**
- No handlebars (not being published for Scala 2.12)
- Update for new annotations APIs
- Do better options parsing

### CODE REVIEW

**Comment by grebe** [comment]
The BOOM test is failing because of assignment to `VOID`. Could someone who knows what's going on help me with that test?

Also, I've updated options parsing to simply extend the firrtl options parser. I want to do this so you can invoke barstools exactly like firrtl. If someone could provide an example fir + anno and a barstool command doing something interesting, I would like to test that my options parsing + updated annotations handling actually works.

**Comment by donggyukim** [comment]
Is it similar to the error in #34?

**Comment by jackkoenig** [comment]
Assignment to `VOID` as in it's failing in FIRRTL? Possibly a manifestation of https://github.com/freechipsproject/firrtl/issues/852? (or is something just straight up not initialized?)

For options parsing, Treadle does this: https://github.com/freechipsproject/treadle/blob/b8660e87946b1d5b352462310afa5eb75908e8ca/src/main/scala/treadle/Driver.scala.

@seldridge is working on making it better, might be worth checking with him. @chick as well.

**Comment by grebe** [comment]
Yes, it's like #34 (number 2 with all the VOIDs). I think it's just an old version of BOOM before all the initialization checks happened.

**Comment by jwright6323** [comment]
@edwardcwang I think we need to settle on one format, but I don't think conf is the right one long-term, either. IMO we need a more verbose output from --repl-seq-mem (or a --repl-seq-mem-mdf e.g.)

**Comment by edwardcwang** [comment]
For now how about we just have both options and when we figure out the "one format to rule the all", we can add an option for that? I don't think an extra option is going to hurt anybody here.

**Comment by jwright6323** [comment]
I am OK with that. I'll add the old option back in.

File: macros/src/main/scala/MacroCompiler.scala [hunk 1]
```diff
@@ -697,7 +702,8 @@ object MacroCompiler extends App {
     .map { case (_, cmd, description) => s"    $cmd: $description" }
   val usage: String = (Seq(
     "Options:",
-    "  -m, --macro-conf: The set of macros to compile in firrtl-generated conf format",
+    "  -m, --macro-conf: The set of macros to compile in firrtl-generated conf format (exclusive with -j)",
+    "  -j, --macro-mdf: The set of macros to compile in MDF JSON format (exclusive with -m)",
```
Comment (edwardcwang) [inline macros/src/main/scala/MacroCompiler.scala]: Does `-c, --macro-conf` and `-m, --macro-mdf` work by any chance?

Comment (jwright6323) [reply macros/src/main/scala/MacroCompiler.scala]: `-c` is taken

File: tapeout/src/main/scala/transforms/Generate.scala [hunk 2]
```diff
@@ -10,208 +10,115 @@ import firrtl.annotations.AnnotationYamlProtocol._
...
-        )
-      ) ++ annotationArray)
-    } else { AnnotationMap(Seq.empty) }
-  }
 
-  private def getSecondPhasePasses: Seq[Transform] = {
-    // always the same for now
+  private def harnessTransforms: Seq[Transform] = {
+    // XXX this is a hack, we really should be checking the masters to see if they are ExtModules
+    val externals = Set(harnessTop.get, synTop.get, "SimSerial")
```
Comment (colinschmidt) [inline tapeout/src/main/scala/transforms/Generate.scala]: Does this make this more brittle than before?
If so what happens when it fails? Is the error message reasonable?

File: tapeout/src/main/scala/transforms/Generate.scala [hunk 3]
```diff
@@ -10,208 +10,115 @@ import firrtl.annotations.AnnotationYamlProtocol._
...
-      Parser.UseInfo,
-      getSecondPhasePasses,
-      getSecondPhaseAnnotations
+  protected def executeHarness: Unit = {
+
+    optionsManager.firrtlOptions = optionsManager.firrtlOptions.copy(
+      customTransforms = harnessTransforms
     )
+
+    firrtl.Driver.execute(optionsManager)
```
Comment (colinschmidt) [inline tapeout/src/main/scala/transforms/Generate.scala]: Is there an issue tracking the double firrtl execution?

File: tapeout/src/main/scala/transforms/RemoveUnusedModules.scala [hunk 4]
```diff
@@ -5,13 +5,15 @@ package barstools.tapeout.transforms
 import firrtl._
 import firrtl.ir._
 import firrtl.passes.Pass
+import firrtl.annotations.{SingleTargetAnnotation, Annotation}
+import firrtl.transforms.DontTouchAnnotation
 
-// Removes all the unused modules in a circuit by recursing through every
-// instance (starting at the main module)
```
Comment (colinschmidt) [inline tapeout/src/main/scala/transforms/RemoveUnusedModules.scala]: Why remove the comment? :(

File: tapeout/src/main/scala/transforms/RemoveUnusedModules.scala [hunk 5]
```diff
@@ -39,21 +41,23 @@ class RemoveUnusedModulesPass extends Pass {
...
-  def outputForm = MidForm
-  def transforms = Seq(new RemoveUnusedModulesPass)
+    //state.circuit.modules.filterNot { usedModuleSet contains _.name } foreach { x => renames.record(ModuleTarget(state.circuit.main, x.name), Seq()) }
 
-  def execute(state: CircuitState): CircuitState = {
-    val ret = runTransforms(state)
-    CircuitState(ret.circuit, outputForm, ret.annotations, ret.renames)
+    val newCircuit = Circuit(state.circuit.info, usedModuleSeq, state.circuit.main)
+    val newAnnos = AnnotationSeq(state.annotations.toSeq.filter { _ match {
+      // XXX This is wrong, but it works for now
```
Comment (colinschmidt) [inline tapeout/src/main/scala/transforms/RemoveUnusedModules.scala]: Point to tracking issue

File: tapeout/src/main/scala/transforms/RemoveUnusedModules.scala [hunk 6]
```diff
@@ -39,21 +41,23 @@ class RemoveUnusedModulesPass extends Pass {
...
-    Circuit(c.info, usedModuleSeq, c.main)
-  }
-}
+    val renames = state.renames.getOrElse(RenameMap())
 
-class RemoveUnusedModules extends Transform with SeqTransformBased {
-  def inputForm = MidForm
-  def outputForm = MidForm
-  def transforms = Seq(new RemoveUnusedModulesPass)
+    //state.circuit.modules.filterNot { usedModuleSet contains _.name } foreach { x => renames.record(ModuleTarget(state.circuit.main, x.name), Seq()) }
```
Comment (colinschmidt) [inline tapeout/src/main/scala/transforms/RemoveUnusedModules.scala]: Note this should be the correct solution except for hack below.

File: tapeout/src/test/scala/transforms/ResetInverterSpec.scala [hunk 7]
```diff
@@ -13,6 +13,8 @@ class ExampleModuleNeedsResetInverted extends Module with ResetInverter {
 
   val r = RegInit(0.U)
 
+  io.out := r
```
Comment (colinschmidt) [inline tapeout/src/test/scala/transforms/ResetInverterSpec.scala]: :laughing:

File: macros/src/main/scala/MacroCompiler.scala [hunk 8]
```diff
@@ -746,25 +755,30 @@ object MacroCompiler extends App {
   def run(args: List[String]) {
     val (params, costParams, forcedMemories) = parseArgs(Map[MacroParam, String](), Map[String, String](), (Set.empty, Set.empty), args)
     try {
-      val macros = Utils.filterForSRAM(mdf.macrolib.Utils.readMDFFromPath(params.get(Macros))).get map (x => (new Macro(x)).blackbox)
+      val macros = if (params.get(MacrosFormat) == Some("conf")) {
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply macros/src/main/scala/MacroCompiler.scala]: done

File: macros/src/main/scala/MacroCompiler.scala [hunk 9]
```diff
@@ -746,25 +755,30 @@ object MacroCompiler extends App {
...
+              params.get(Macros).get, params.get(MacrosFormat), params.get(Library),
               CostMetric.getCostMetric(params.getOrElse(CostFunc, "default"), costParams),
               MacroCompilerAnnotation.stringToCompilerMode(params.getOrElse(Mode, "default")),
               params.contains(UseCompiler),
               forceCompile = forcedMemories._1, forceSynflops = forcedMemories._2
             )
           ))
         )
-        val state = CircuitState(circuit, HighForm, Some(annotations))
+        // Append a NoDCEAnnotation to avoid dead code elimination removing the non-parent SRAMs
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply macros/src/main/scala/MacroCompiler.scala]: Yes, but this is not where it should go (this code is correct). The problem is in the tapeout pass.

File: macros/src/main/scala/MemConf.scala [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+package barstools.macros
+
+import scala.util.matching._
+
+sealed abstract class MemPort(val name: String) { override def toString = name }
+
+case object ReadPort extends MemPort("read")
+case object WritePort extends MemPort("write")
+case object MaskWritePort extends MemPort("mwrite")
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply macros/src/main/scala/MemConf.scala]: Done

File: macros/src/main/scala/MemConf.scala [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+case object ReadWritePort extends MemPort("rw")
+case object MaskReadWritePort extends MemPort("mrw")
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply macros/src/main/scala/MemConf.scala]: done

File: tapeout/src/main/scala/transforms/Generate.scala (hunk 3 above)
Comment (jwright6323) [reply tapeout/src/main/scala/transforms/Generate.scala]: #37

File: tapeout/src/main/scala/transforms/RemoveUnusedModules.scala (hunk 4 above)
Comment (jwright6323) [reply tapeout/src/main/scala/transforms/RemoveUnusedModules.scala]: Accident, restored.

File: tapeout/src/main/scala/transforms/RemoveUnusedModules.scala (hunk 5 above)
Comment (jwright6323) [reply tapeout/src/main/scala/transforms/RemoveUnusedModules.scala]: Done

File: tapeout/src/main/scala/transforms/RemoveUnusedModules.scala (hunk 6 above)
Comment (jwright6323) [reply tapeout/src/main/scala/transforms/RemoveUnusedModules.scala]: done

File: tapeout/src/test/scala/transforms/ResetInverterSpec.scala (hunk 7 above)
Comment (jwright6323) [reply tapeout/src/test/scala/transforms/ResetInverterSpec.scala]: no action

File: macros/src/main/scala/MacroCompiler.scala [hunk 12]
```diff
@@ -717,8 +724,10 @@ object MacroCompiler extends App {
                 args: List[String]): (MacroParamMap, CostParamMap, ForcedMemories) =
     args match {
       case Nil => (map, costMap, forcedMemories)
-      case ("-m" | "--macro-list") :: value :: tail =>
-        parseArgs(map + (Macros  -> value), costMap, forcedMemories, tail)
+      case ("-n" | "--macro-conf") :: value :: tail =>
+        parseArgs(map + (Macros  -> value) + (MacrosFormat -> "conf"), costMap, forcedMemories, tail)
+      case ("-m" | "--macro-mdf") :: value :: tail =>
+        parseArgs(map + (Macros  -> value) + (MacrosFormat -> "mdf"), costMap, forcedMemories, tail)
```
[the remark this answers is not in the document]
Comment (jwright6323) [reply macros/src/main/scala/MacroCompiler.scala]: It will take the last option specified. I think this is intuitive enough behavior to leave it as is.

**Pull request merged**